Benchmarking
***********************

To assess the general applicability of trimAl and establish empirical heuristics for automatic parameter selection, we conducted a benchmark analysis.

Three different sets were used for this benchmark. One of these sets had been previously employed [Talavera G, Castresana J., 2007] to evaluate the enhancement in phylogenetic performance post-alignment trimming. This set involved evolutionary simulations of protein sequences with various lengths (400 to 3200 positions) generated using ROSE [Stoye J et al, 1998]. The simulations were conducted along phylogenetic trees with 16 tips, featuring three different topologies exhibiting varying levels of symmetry. The branch lengths of these trees were multiplied by 0.5, 1, and 2, resulting in six distinct phylogenetic trees.

The other two sets of alignments were newly generated by us to expand the dataset to cases with 32 and 64 tree tips, which are considered more realistic for phylogenomic analyses. To create these additional sets, we took reference trees from the previous study and, using ETE [http://etetoolkit.org], generated twelve new reference trees. Six of them had 32 tips, and the other six had 64 tips, maintaining the same level of asymmetry as those in the Talavera et al. study. The lengths of their branches were multiplied to achieve the same three levels of divergence (0.5, 1, and 2.0) as the previous study.

These reference trees were employed to generate corresponding sets of alignments, following the approach outlined in Talavera et al. (2007). We utilized the ROSE v1.3 program [Stoye J et al, 1998] with the same seed protein and parameters described in [Talavera G, Castresana J., 2007] to generate benchmark sets. The simulations included insertions and deletions with a probability of 0.03. Other parameters for the simulation were also the same described in [Talavera G, Castresana J., 2007]. The strategy for inferring the patterns of rate heterogeneity of the seed proteins was also identical. Similar to the set of 16 sequences, the sets generated by ROSE contained simulated protein sequences of various lengths (400 to 3200 positions) and different topologies.


.. table:: A summary of the total number of alignments used in our benchmark sorted in different groups depending on the number of sequences, kind of topology, tree divergence and average length of the sequences before being aligned. The last row presents the overall count of alignments for each sequence length.
   :name: benchmark-table
   :widths: 10, 15, 25, 10, 10, 10, 10, 10
   :class: longtable
   :align: center
   :width: 66%

   +-----------------------+----------------+---------------------+--------------------------------+
   | Number of sequences   | Tree topology  | Sequence divergence | Alignment length               |
   +=======================+================+=====================+=====+=====+======+======+======+
   |                       |                |                     | 400 | 800 | 1200 | 1600 | 3200 |
   +-----------------------+----------------+---------------------+-----+-----+------+------+------+
   | 16                    | Asymmetric     | 0.5                 | 300 | 300 | 300  | 300  | 300  |
   |                       |                +---------------------+-----+-----+------+------+------+
   |                       |                | 1                   | 300 | 300 | 300  | 300  | 300  |
   |                       |                +---------------------+-----+-----+------+------+------+
   |                       |                | 2                   | 300 | 300 | 300  | 300  | 300  |
   |                       +----------------+---------------------+-----+-----+------+------+------+
   |                       | Symmetric      | 0.5                 | 300 | 300 | 300  | 300  | 300  |
   |                       |                +---------------------+-----+-----+------+------+------+
   |                       |                | 1                   | 300 | 300 | 300  | 300  | 300  |
   |                       |                +---------------------+-----+-----+------+------+------+
   |                       |                | 2                   | 300 | 300 | 300  | 300  | 300  |
   +-----------------------+----------------+---------------------+-----+-----+------+------+------+
   | 32                    | Asymmetric     | 0.5                 | 100 | 100 | 100  | 100  | 100  |
   |                       |                +---------------------+-----+-----+------+------+------+
   |                       |                | 1                   | 100 | 100 | 100  | 100  | 100  |
   |                       |                +---------------------+-----+-----+------+------+------+
   |                       |                | 2                   | 100 | 100 | 100  | 100  | 100  |
   |                       +----------------+---------------------+-----+-----+------+------+------+
   |                       | Symmetric      | 0.5                 | 100 | 100 | 100  | 100  | 100  |
   |                       |                +---------------------+-----+-----+------+------+------+
   |                       |                | 1                   | 100 | 100 | 100  | 100  | 100  |
   |                       |                +---------------------+-----+-----+------+------+------+
   |                       |                | 2                   | 100 | 100 | 100  | 100  | 100  |
   +-----------------------+----------------+---------------------+-----+-----+------+------+------+
   | 64                    | Asymmetric     | 0.5                 | 100 | 100 | 100  | 100  | 100  |
   |                       |                +---------------------+-----+-----+------+------+------+
   |                       |                | 1                   | 100 | 100 | 100  | 100  | 100  |
   |                       |                +---------------------+-----+-----+------+------+------+
   |                       |                | 2                   | 100 | 100 | 100  | 100  | 100  |
   |                       +----------------+---------------------+-----+-----+------+------+------+
   |                       | Symmetric      | 0.5                 | 100 | 100 | 100  | 100  | 100  |
   |                       |                +---------------------+-----+-----+------+------+------+
   |                       |                | 1                   | 100 | 100 | 100  | 100  | 100  |
   |                       |                +---------------------+-----+-----+------+------+------+
   |                       |                | 2                   | 100 | 100 | 100  | 100  | 100  |
   +-----------------------+----------------+---------------------+-----+-----+------+------+------+
   | Total number of alignments                                   | 3000| 3000| 3000 | 3000 | 3000 |
   +-----------------------+----------------+---------------------+-----+-----+------+------+------+

